unit/solvers/refinement/string_refinement/union_find_replace.cpp

```diff
@@ -0,0 +1,68 @@
+/*******************************************************************\
+
+ Module: Unit tests for union_find_replacet in
+   solvers/refinement/string_refinement.cpp
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#include <testing-utils/catch.hpp>
+
+#include <util/arith_tools.h>
+#include <util/c_types.h>
+#include <util/std_types.h>
+#include <util/std_expr.h>
+#include <solvers/refinement/string_refinement.h>
+
+SCENARIO("union_find_replace",
+  "[core][solvers][refinement][string_refinement]")
+{
+  GIVEN("An empty dictionary")
+  {
+    union_find_replacet dict;
+    pointer_typet char_pointer_type=pointer_type(unsignedbv_typet(16));
+    const symbol_exprt a("a", char_pointer_type);
+    const symbol_exprt b("b", char_pointer_type);
+    const symbol_exprt c("c", char_pointer_type);
+    const symbol_exprt d("d", char_pointer_type);
+    const symbol_exprt e("e", char_pointer_type);
+    const symbol_exprt f("f", char_pointer_type);
+
+    WHEN("Relations a=b, a=c, d=b, e=f are added")
+    {
+      dict.add_symbol(a, b);
+      dict.add_symbol(a, c);
+      dict.add_symbol(d, b);
+      dict.add_symbol(e, f);
+      THEN("find(d)=find(c), but find(e)!=find(a)")
+      {
+        REQUIRE(dict.find_expr(d)==dict.find_expr(c));
+        REQUIRE(dict.find_expr(e)!=dict.find_expr(a));
+      }
+
+      GIVEN("Expressions a+e, c+f, c+d")
+      {
+        plus_exprt a_plus_e(a, e);
+        plus_exprt c_plus_f(c, f);
+        plus_exprt c_plus_d(c, d);
+        WHEN("We use the dictionary for replacement")
+        {
+          dict.replace_expr(a_plus_e);
+          dict.replace_expr(c_plus_f);
+          dict.replace_expr(c_plus_d);
+          THEN("a+e=c+f but a+e!=c+d")
+          {
+            REQUIRE(a_plus_e==c_plus_f);
+            REQUIRE(a_plus_e!=c_plus_d);
+          }
+        }
+      }
+
+      THEN("Introducing cycles does not cause infinite loops or exceptions")
+      {
+        dict.add_symbol(c, d);
```

Comment:
There should probably be some kind of checks here that the `dict` is still reasonable?

Reply:
added a couple of check after this (one positive and one negative)